_posts/2018-12-04-imposing-invariants-in-scala.md

```diff
@@ -0,0 +1,359 @@
+---
+layout: post
+author: cmteixeira
+title: Imposing invariants in Scala
+---
+
+
+
+Imagine a opera house allows reservations to a show. For each reservation two people can come: The buyer of the reservation himself, and his guest.  
+As in any show-house, there are many types of seats; some are standing, some normal and some on vip tribunes.
+
+
+
+As a restriction from the opera house, both buyer and guest must be on the same seating area. For example, it is not permitted for one ticket to be standing while the other for the normal seated area.  
+I believe this is what people call an invariant. A requirement that must hold for the data to have meaning.  
+How to enforce then this constraint ?  
+Let us first define the data types relevant to the problem:  
+
+
+```scala
+sealed trait Seat
+
+case class NormalSeat(row: Int, column: Int) extends Seat
+case class VipTribune(tribuneId: String) extends Seat
+case object Standing extends Seat
+```
+
+In what follows it will be useful to discern the type relation above on the following diagram.
```

Comment:
I understand what you mean here, but the sentence has a weird construction in English. It makes sense in Portuguese.
Right now, I don't have any suggestion. I'll come back to this later.

Reply:
I agree with Sofia, I believe a better way to express this would be `This DSL might be easier to understand with the following diagram:`

Reply:
It's not the verb, it's the sentence construction. Perhaps this:
> The following diagram is useful for discerning the relation above and understanding the explanations that follow.

Does this seem good?

Reply:
Perfect. Will incorporate.

Reply:
Incorporated. Thanks